minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,823 @@
+
+
+ZShape = namedtuple("ZShape", ["m", "t"], defaults=(None, None))
+
+
+
+class EdgeKind(Enum):
+    INTERNAL = 1
+    EXTERNAL = 2
+    ODD = 3
+
+
+class NodeKind(Enum):
+    VERTICAL = 0
+    HORIZONTAL = 1
+
+
+class Edge:
+    """Represents an edge of a graph in a canonical order.
+
+    Args:
+        x (Any): One endpoint of edge.
+        y (Any): Another endpoint of edge.
+
+    ..note:: ``x`` and ``y`` must be mutually comparable.
+    """
+
+    def __init__(self, x: Any, y: Any) -> None:
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x: Any, y: Any):
+        """Returns ordered tuple corresponding to the set {x, y}."""
+        if x < y:
+            return (x, y)
+        else:
+            return (y, x)
+
+    def __hash__(self) -> int:
+        return hash(self._edge)
+
+    def __getitem__(self, index: int) -> int:
+        return self._edge[index]
+
+    def __eq__(self, other: Edge) -> bool:
+        return self._edge == other._edge
+
+    def __str__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+
+class ZEdge(Edge):
+    """Represents an edge in a graph with Zephyr topology.
+
+    Args:
+        x (ZNode): Endpoint of edge. Must have same shape as ``y``.
+        y (ZNode): Endpoint of edge. Must have same shape as ``x``
+        check_edge_valid (bool, optional): Flag to whether check the validity of values and types of ``x``, ``y``.
+            Defaults to True.
+
+    Raises:
+        TypeError: If either of x or y is not :class:`ZNode`.
+        ValueError: If x, y do not have the same shape.
+        ValueError: If x, y are not neighbors in a perfect yield (quotient)
+            Zephyr graph.
+
+    Example 1:
+    >>> from minorminer.utils.zephyr.node_edge import ZNode, ZEdge
+    >>> e = ZEdge(ZNode((3, 2)), ZNode((7, 2)))
+    >>> print(e)
+    ZEdge(ZNode(CartesianCoord(x=3, y=2, k=None)), ZNode(CartesianCoord(x=7, y=2, k=None)))
+    Example 2:
+    >>> from minorminer.utils.zephyr.node_edge import ZNode, ZEdge
+    >>> ZEdge(ZNode((2, 3)), ZNode((6, 3))) # raises error, since the two are not neighbors
+    """
+
+    def __init__(
+        self,
+        x: ZNode,
+        y: ZNode,
+        check_edge_valid: bool = True,
+    ) -> None:
+        if check_edge_valid:
+            if not isinstance(x, ZNode) or not isinstance(y, ZNode):
+                raise TypeError(f"Expected x, y to be ZNode, got {type(x), type(y)}")
+
+            if x.shape != y.shape:
+                raise ValueError(f"Expected x, y to have the same shape, got {x.shape, y.shape}")
+
+            for kind in EdgeKind:
+                if x.is_neighbor(y, nbr_kind=kind):
+                    self._edge_kind = kind
+                    break
+            else:
+                raise ValueError(f"Expected x, y to be neighbors, got {x, y}")
```

Comment:
Since you already do practically the same check in `edge_kind` property, you could simplify this to:

```suggestion
            # edge_kind property calculation verifies edge is a valid Zephyr edge
            self.edge_kind
```

If you apply this, you'll also need to move the init above.